### PR TITLE
Add Data responses and TypedDataResponse

### PR DESCRIPTION
**Pull request by aydenp**
In some cases, you may want to just return raw `Data` as a Response. Before this commit, you might have had to make your own `ResponseEncodable` type that just provides the data, which seems a bit much for such a simple function.

Say we have this super important data that we need to respond with:

    let data = "Hello!".data(using: .utf8)!

At it's simplest, this new functionality allows you to respond with a `Data` object by simply returning it:

    router.get("hello") { req in
        return data
    }

However, sometimes, you might want to return this data with a custom content type. In this case, you can simply get a `TypedDataResponse` from a new Data extension, and return it, instead of the `Data`:

    router.get("hello") { req in
        return data.response(type: .html)
    }

As an aside, `TypedDataResponse` is also used under the hood if you just return `Data` on it's own, with a content type of `MediaType.any`.

### Checklist

- [x] Circle CI is passing (code compiles and passes tests).
- [x] There are no breaking changes to public API.
- [x] New test cases have been added where appropriate.
- [x] All new code has been commented with doc blocks `///`.

### CODE REVIEW

File: Sources/Vapor/Response/TypedDataResponse.swift [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/// A container representing `Data` with a specific content type (used for returning as a `Response`).
+public struct TypedDataResponse: ResponseEncodable {
```
Comment (tanner0101) [inline Sources/Vapor/Response/TypedDataResponse.swift]: We should make this type private and just publicize `ResponseEncodable`.

File: Sources/Vapor/Response/TypedDataResponse.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    /// See `ResponseEncodable`.
+    public func encode(for req: Request) throws -> EventLoopFuture<Response> {
+        return try response(type: .any).encode(for: req)
+    }
+    
+    /**
+     Get a `ResponseEncodable` container holding this data with your provided contentType.
+     - parameter contentType: The type of data to return the container with.
+     */
+    public func response(type: MediaType) -> TypedDataResponse {
```
Comment (tanner0101) [inline Sources/Vapor/Response/TypedDataResponse.swift]: Here we should be able to just return `ResponseEncodable`.

File: Sources/Vapor/Response/TypedDataResponse.swift [hunk 3]
```diff
@@ -0,0 +1,30 @@
+/// A container representing `Data` with a specific content type (used for returning as a `Response`).
+public struct TypedDataResponse: ResponseEncodable {
+    /// The data held by this container.
+    let data: Data
+    /// The type of the data held by this container.
```
Comment (tanner0101) [inline Sources/Vapor/Response/TypedDataResponse.swift]: should be a newline above this comment.

File: Sources/Vapor/Response/TypedDataResponse.swift [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+extension Data: ResponseEncodable {
+    /// See `ResponseEncodable`.
+    public func encode(for req: Request) throws -> EventLoopFuture<Response> {
+        return try response(type: .any).encode(for: req)
+    }
+    
+    /**
+     Get a `ResponseEncodable` container holding this data with your provided contentType.
```
Comment (tanner0101) [inline Sources/Vapor/Response/TypedDataResponse.swift]: couple gardening notes here:
- vapor uses `///` format comments
- `- parameter`s should be listed as a separate bulleted list, see any of the other doc blocks for an example
- methods expected to be used directly by the end user should have a small example code section if possible! that really helps people figure out how to use it.

File: Sources/Vapor/Response/TypedDataResponse.swift (hunk 1 above)
Comment (aydenp) [reply Sources/Vapor/Response/TypedDataResponse.swift]: `ResponseEncodable` is already public, but I agree it that this type should be private. An earlier iteration of it didn't have the `Data` extension for types, so I'd forgotten to make it private.

**Comment by penny-coin** [comment]
Hey @aydenp, you just merged a pull request, have a coin! 

You now have 1 coins.

**Comment by tanner0101** [comment]
Ended on something slightly different, but I think it better matches the other available APIs which will make things more intuitive.

```swift
// default to plaintext
router.get("hello") { req in
    return req.makeResponse("Hello!")
}

// specify content type
router.get("hello-html") { req in
    return req.makeResponse("Hey!", as: .html)
}

// and of course
router.get("hello-json") { req in
    let res = req.makeResponse()
    try res.content.encode(json: ["hello": "world"], as: .json)
    return res
}
```

**Comment by tanner0101** [comment]
Worth noting that both `String` and `Data` conform to `HTTPBodyLosslessRepresentable`. Any other types that conform can also be used with this new method.

**Comment by nullpixel** [comment]
Do your changes let you just return `Data`?

**Comment by tanner0101** [comment]
@nullpixel what Content-Type would `Data` use? There wasn't an obvious best choice there which is why I omitted that part.

**Comment by nullpixel** [comment]
Well, `.any` makes the most sense, since, well, it can be anything.

**Pull request merged**